### PR TITLE
Initial System.Net.Http 4.1 contract

### PR DESCRIPTION
**Pull request by davidsh**
To support important customer scenarios for CoreFx RTM, we will be updating the System.Net.Http contract from 4.0 to 4.1. We are focusing on HttpClientHandler enhancements.

We will be submitting the final proposed list of additions as part of issue #7623. As part of updating the contract to 4.1, we are also OOB'ing the Desktop (net46 Windows) implementation so that the package won't be a facade over Desktop but rather a full implementation. This is possible since System.Net.Http has always been a standalone dll in the .NET Framework. As part of this, we are changing the implementation for Desktop from the legacy HttpWebRequest sockets stack to the WinHttpHandler stack based on native WinHTTP. This will give us a good path forward as WinHttpHandler stack will be getting HTTP/2.0 support with newer versions of Windows 10.

This initial PR has a few API additions that we are definitely going to be adding. It's important to get these changes into the builds so that we can fully bake the OOB Http.dll assembly on all the platforms and continue package testing.

There is an existing System.Net.Http.WebRequestHandler class in the .NET Framework located in System.Net.Http.WebRequest.dll. This class is a subclass of HttpClientHandler. Due to this, we are being careful not to cause collisions with some of these added properties. WebRequestHandler has a 'ClientCertificates' property and a 'ServerCertificateValidationCallback' property. So, we are aligning/renaming the properties as we add them to HttpClientHandler OOB assembly. This may get altered slightly as we finalize the 4.1 System.Net.Http API.

### CODE REVIEW

**Comment by davidsh** [comment]
cc: @stephentoub @CIPop @ericstj @weshaggard @joshfree @SidharthNabar

File: src/System.Net.Http/ref/System.Net.Http.cs [hunk 1]
```diff
@@ -80,13 +80,16 @@ public partial class HttpClientHandler : System.Net.Http.HttpMessageHandler
         public HttpClientHandler() { }
         public bool AllowAutoRedirect { get { return default(bool); } set { } }
         public System.Net.DecompressionMethods AutomaticDecompression { get { return default(System.Net.DecompressionMethods); } set { } }
+        public bool CheckCertificateRevocationList { get { return default(bool); } set { } }
         public System.Net.Http.ClientCertificateOption ClientCertificateOptions { get { return default(System.Net.Http.ClientCertificateOption); } set { } }
+        public System.Security.Cryptography.X509Certificates.X509CertificateCollection ClientCertificates { get { return default(System.Security.Cryptography.X509Certificates.X509Certificate2Collection); } }
```
Comment (stephentoub) [inline src/System.Net.Http/ref/System.Net.Http.cs]: If the return type is X509CertificateCollection, the return statement should probably be returning default(X509CertificateCollection) rather than default(X509Certificate2Collection).

File: src/System.Net.Http/src/System/Net/Http/HttpContent.cs [hunk 2]
```diff
@@ -139,12 +139,13 @@ internal void SetBuffer(byte[] buffer, int offset, int count)
         {
             _bufferedContent = new MemoryStream(buffer, offset, count, writable: false, publiclyVisible: true);
         }
-        
+
         internal bool TryGetBuffer(out ArraySegment<byte> buffer)
         {
+            buffer = default(ArraySegment<byte>);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/HttpContent.cs]: Is this necessary to get it to compile?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs [hunk 3]
```diff
@@ -261,12 +261,16 @@ public static IEnumerable<object[]> UseCallback_ValidCertificate_ExpectedValuesD
         }
 
         private static bool BackendSupportsCustomCertificateHandling =>
-            RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ||
+#if unix
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: We've tried hard to avoid platform-based ifdefs like this.  Why is this necessary?

File: src/System.Net.Http/ref/System.Net.Http.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http/ref/System.Net.Http.cs]: Ok. But the actual property will return an X509Certificate2Collection (which is derived from X509CertificateCollection). We would want the API to return the derived type but we would have a type conflict with a similarly named property in WebRequestHandler.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs (hunk 3 above)
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: This code was in the Unix only compile of the tests you added. Without the 'if', it will end up p/invoking into a unix API.  Won't that fail?

File: src/System.Net.Http/src/System/Net/Http/HttpContent.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/HttpContent.cs]: @ericstj had added that line when I was compiling a separate 'net46' OOB implementation against mscorlib, system. So, I left it in assuming it was needed.

But now that we are using a single OOB implementation and compiling against system.runtime, it isn't necessary. I'll remove it.

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs (hunk 3 above)
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: Actually, I just discovered that by leaving this line in:

```c#
RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ||
```

it will avoid calling into the p/invoke since the expression will be true.  So, yes, I can remove the 'if' def logic.

File: src/System.Net.Http/ref/System.Net.Http.cs (hunk 1 above)
Comment (stephentoub) [inline src/System.Net.Http/ref/System.Net.Http.cs]: What's the rationale for having the API defined with a return type of ```X509CertificateCollection``` rather than ```X509Certificate2Collection```?  Does doing so afford us more flexibility in some regard?

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ServerCertificates.cs]: Exactly. Thanks.

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.Http/ref/System.Net.Http.cs (hunk 1 above)
Comment (davidsh) [inline src/System.Net.Http/ref/System.Net.Http.cs]: >What's the rationale for having the API defined with a return type of X509CertificateCollection rather than X509Certificate2Collection? 

The problem is that we have an existing type in the .NET Framework, `WebRequestHandler`. This type lives in a separate DLL, System.Net.Http.WebRequest.dll.  And `WebRequestHandler` already has a property called `ClientCertificates` and it is on type `X509CertificateCollection`.

So, we can't really add a new property to `HttpClientHandler` with the same name of `ClientCertificates` but have it be of type `X509Certifcate2Collection`.  We are concerned about what happens if a developer take the OOB package of System.Net.Http.dll and also adds a reference to System.Net.Http.WebRequest.dll.  There would be conflicts in type of a similarly named property.

File: src/System.Net.Http/ref/project.json [hunk 4]
```diff
@@ -3,14 +3,15 @@
     "System.Runtime": "4.0.0",
     "System.Threading.Tasks": "4.0.0",
     "System.IO": "4.0.0",
-    "System.Net.Primitives": "4.0.0",
+    "System.Net.Primitives": "4.0.10",
+    "System.Security.Cryptography.X509Certificates": "4.1.0-rc3-24008-00",
```
Comment (weshaggard) [inline src/System.Net.Http/ref/project.json]: This is going to cause Crypto to be pulled into standard library (fyi @bartonjs), I'm not necessarily against it but I want others to be aware of it and make sure we are intentional about it.

cc @KrzysztofCwalina @terrajobst

File: src/System.Net.Http/src/netcore50/project.json [hunk 5]
```diff
@@ -16,6 +16,7 @@
         "System.Resources.ResourceManager": "4.0.0",
         "System.Runtime.Extensions": "4.0.10",
         "System.Runtime.WindowsRuntime": "4.0.10",
+        "System.Security.Cryptography.X509Certificates": "4.1.0-rc3-24008-00",
```
Comment (weshaggard) [inline src/System.Net.Http/src/netcore50/project.json]: @bartonjs are you ready for Crypt to be pulled in and supported on UWP?

File: src/System.Net.Http/ref/System.Net.Http.cs (hunk 1 above)
Comment (weshaggard) [inline src/System.Net.Http/ref/System.Net.Http.cs]: As for the different return type of default(X509Certificate2Collection) it really doesn't matter for the contract reference assembly as it that code will never be executed and ultimately the value in both cases is null, so I would personally keep it matching the return type here.

File: src/System.Net.Http/ref/project.json (hunk 4 above)
Comment (davidsh) [inline src/System.Net.Http/ref/project.json]: Is System.Net.Http part of "standard library"?  I'm not sure I understand what that is and what contracts are part of "standard library". But System.Net.Http needs to have Crypto moving forward.

File: src/System.Net.Http/src/netcore50/project.json (hunk 5 above)
Comment (davidsh) [inline src/System.Net.Http/src/netcore50/project.json]: Last time I spoke with @bartonjs about this, Crypto was already supported on UWP but not yet part of the meta-package.

File: src/System.Net.Http/src/System/Net/Http/HttpContent.cs (hunk 2 above)
Comment (ericstj) [inline src/System.Net.Http/src/System/Net/Http/HttpContent.cs]: Was getting a compiler error here when compiling locally with roslyn.  Double check that this builds in a VS2015 command-prompt.

File: src/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj [hunk 6]
```diff
@@ -62,11 +63,13 @@
     <Compile Include="TestHelper.cs" />
     <Compile Include="XunitTestAssemblyAttributes.cs" />
   </ItemGroup>
+  <PropertyGroup Condition=" '$(TargetsUnix)' == 'true' ">
+    <DefineConstants>$(DefineConstants);unix</DefineConstants>
```
Comment (ericstj) [inline src/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: Per @stephentoub's comment you can remove this.

Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/System.Net.Http.Functional.Tests.csproj]: I removed this in the 2nd commit.

File: src/System.Net.Http/src/System/Net/Http/HttpContent.cs (hunk 2 above)
Comment (davidsh) [inline src/System.Net.Http/src/System/Net/Http/HttpContent.cs]: This PR compiles fine in CI, so that line isn't needed and I took it out in the 2nd commit.

File: src/System.Net.Http/ref/4.0.0/System.Net.Http.csproj [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <OutputType>Library</OutputType>
+    <PackageTargetFramework>netstandard1.1</PackageTargetFramework>
+    <NuGetTargetMoniker>.NETStandard,Version=v1.1</NuGetTargetMoniker>
+  </PropertyGroup>
+  <ItemGroup>
+    <Compile Include="System.Net.Http.cs" />
```
Comment (ericstj) [inline src/System.Net.Http/ref/4.0.0/System.Net.Http.csproj]: We should use a depproj for this instead of CSPROJ.  We need to consider 4.0.0 ref assembly frozen.

File: src/System.Net.Http/ref/System.Net.Http.csproj [hunk 8]
```diff
@@ -2,10 +2,11 @@
 <Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <PropertyGroup>
-    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <AssemblyVersion>4.1.0.0</AssemblyVersion>
     <OutputType>Library</OutputType>
-    <PackageTargetFramework>netstandard1.1</PackageTargetFramework>
-    <NuGetTargetMoniker>.NETStandard,Version=v1.1</NuGetTargetMoniker>
+    <PackageTargetFramework Condition="'$(PackageTargetFramework)' == ''">netstandard1.3</PackageTargetFramework>
+    <PackageTargetFramework Condition="'$(TargetGroup)' == 'net46'">net46</PackageTargetFramework>
```
Comment (ericstj) [inline src/System.Net.Http/ref/System.Net.Http.csproj]: This is unnecessary.  Reference CSProj only ever builds for netstandard1.3.

File: src/System.Net.Http/ref/4.0.0/System.Net.Http.cs [hunk 9]
```diff
@@ -0,0 +1,607 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// ------------------------------------------------------------------------------
+// Changes to this file must follow the http://aka.ms/api-review process.
+// ------------------------------------------------------------------------------
+
+
+namespace System.Net.Http
```
Comment (ericstj) [inline src/System.Net.Http/ref/4.0.0/System.Net.Http.cs]: This entire file should be deleted.  Please use a depproj to redist the previously shipped reference assembly for 4.0.0.0

File: src/System.Net.Http/pkg/System.Net.Http.pkgproj [hunk 10]
```diff
@@ -1,19 +1,23 @@
...
-      <SupportedFramework>net45;netcore45;netstandardapp1.5;wpa81</SupportedFramework>
+      <SupportedFramework>net46;netcore50;netstandardapp1.5</SupportedFramework>
     </ProjectReference>
     <ProjectReference Include="..\src\System.Net.Http.builds" />
     <ProjectReference Include="$(NativePackagePath)\runtime.native.System\runtime.native.System.pkgproj" />
     <ProjectReference Include="$(NativePackagePath)\runtime.native.System.Net.Http\runtime.native.System.Net.Http.pkgproj" />
     <ProjectReference Include="$(NativePackagePath)\runtime.native.System.Security.Cryptography\runtime.native.System.Security.Cryptography.pkgproj" />
   </ItemGroup>
   <ItemGroup>
+    <OutOfBoxFramework Include="net46" />
```
Comment (ericstj) [inline src/System.Net.Http/pkg/System.Net.Http.pkgproj]: Please add a comment here.  "We treat net46 as out-of-band, using the PCL reference assembly and PCL implementation."

Also, we might need something to avoid the `frameworkAssembly` reference from net45 applying in a net46 app.  That can be addressed separately.

File: src/System.Net.Http/ref/4.0.0/System.Net.Http.csproj (hunk 7 above)
Comment (davidsh) [inline src/System.Net.Http/ref/4.0.0/System.Net.Http.csproj]: Changed to depproj.

File: src/System.Net.Http/ref/System.Net.Http.csproj (hunk 8 above)
Comment (davidsh) [inline src/System.Net.Http/ref/System.Net.Http.csproj]: Removed 'net46' line.

File: src/System.Net.Http/ref/4.0.0/System.Net.Http.cs (hunk 9 above)
Comment (davidsh) [inline src/System.Net.Http/ref/4.0.0/System.Net.Http.cs]: Removed .cs file and using .depproj.

File: src/System.Net.Http/pkg/System.Net.Http.pkgproj (hunk 10 above)
Comment (davidsh) [inline src/System.Net.Http/pkg/System.Net.Http.pkgproj]: >Please add a comment here. "We treat net46 as out-of-band, using the PCL reference assembly and PCL implementation."

Added comment.

File: src/System.Net.Http/src/netcore50/project.json (hunk 5 above)
Comment (bartonjs) [inline src/System.Net.Http/src/netcore50/project.json]: * In theory, everything works.
* In theory, we run the tests (we had a failure in the ToF tree for UWP because of UTC vs PDT host issues, so I'm pretty sure we run them)

But I don't know that anyone has created a new UWP project, referenced the crypto packages, written a program that used them, and executed it on a phone.  Doing that is definitely on my TODO list, but other higher urgency work keeps slipping in front of it.

Comment (davidsh) [inline src/System.Net.Http/src/netcore50/project.json]: @bartonjs So, what is the next step? Are we blocked from rev'ing the System.Net.Http contract? Or do we just do it and work out the problems as we go along?

Comment (bartonjs) [inline src/System.Net.Http/src/netcore50/project.json]: @davidsh As long as we're not hoping to rush marking 4.1 as stable I'm good with "work out the problems as we go along".  The crypto packages are going to be stable by RTM, and that includes UWP unless we do a huge amount of gymnastics.

Comment (davidsh) [inline src/System.Net.Http/src/netcore50/project.json]: The point was to make 4.1 stable and ship for RTM.

@stephentoub How do you want to proceed? We cannot update the contract to 4.1 and ship (mark stable) for RTM unless we get Crypto to support UWP for netstandard1.3.

File: src/System.Net.Http/pkg/System.Net.Http.pkgproj [hunk 11]
```diff
@@ -1,19 +1,25 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <Version>4.0.1.0</Version>
+    <Version>4.1.0.0</Version>
   </PropertyGroup>
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
   <ItemGroup>
+    <ProjectReference Include="..\ref\4.0.0\System.Net.Http.depproj">
```
Comment (ericstj) [inline src/System.Net.Http/pkg/System.Net.Http.pkgproj]: You can use `4.0` for the folder name since this one doesn't have any old 4.0.n0 contract versions.

File: src/System.Net.Http/src/netcore50/project.json (hunk 5 above)
Comment (stephentoub) [inline src/System.Net.Http/src/netcore50/project.json]: @davidsh, I believe @bartonjs is saying it's fine to take this dependency for RTM.  @bartonjs, correct?

Comment (bartonjs) [inline src/System.Net.Http/src/netcore50/project.json]: Correct.   I just don't want to commit to marking crypto stable before RTM.

Comment (davidsh) [inline src/System.Net.Http/src/netcore50/project.json]: RTM is fine. We don't plan to mark any revised System.Net.* packages as stable until RTM.

File: src/System.Net.Http/pkg/System.Net.Http.pkgproj (hunk 11 above)
Comment (davidsh) [inline src/System.Net.Http/pkg/System.Net.Http.pkgproj]: Ok. I'll change the folder name to `4.0` for the next commit.

File: src/System.Net.Http/ref/project.json (hunk 4 above)
Comment (weshaggard) [inline src/System.Net.Http/ref/project.json]: Yes System.Net.Http is part of standard library. The easiest way to think about standard library is to see what is in NETStandard.Library (https://github.com/dotnet/corefx/blob/master/pkg/NETStandard.Library/NETStandard.Library.packages.targets) meta-package which we are representation we are using for standard library. I know today the implementation depends on Crypto but now after this change the public API surface will as well. At any rate I chatted with @KrzysztofCwalina offline and I think we are OK adding Crypto to standard library.

Can you please add System.Security.Cryptography.X509Certificates to the list at https://github.com/dotnet/corefx/blob/master/pkg/NETStandard.Library/NETStandard.Library.packages.targets?

Comment (davidsh) [inline src/System.Net.Http/ref/project.json]: Ok.  I'll add it to the list for NETStandard.Library.

Comment (ericstj) [inline src/System.Net.Http/ref/project.json]: This will actually add its compile-time closure to NETStandard.Library as well: X509Certificates, System.Security.Cryptography.Algorithms, System.Security.Cryptography.Encoding, System.Security.Cryptography.Primitives

Comment (davidsh) [inline src/System.Net.Http/ref/project.json]: Does that mean I should add these other assemblies to NETStandard.Library targets file:

* System.Security.Cryptography.Algorithms
* System.Security.Cryptography.Encoding
* System.Security.Cryptography.Primitives

**Comment by davidsh** [comment]
Thx for the feedback so far. I'd like to get this merged today so we can get it into the builds and test out  more package scenarios.

File: src/System.Net.Http/ref/project.json (hunk 4 above)
Comment (weshaggard) [inline src/System.Net.Http/ref/project.json]: Yes please.

Comment (davidsh) [inline src/System.Net.Http/ref/project.json]: Ok. I'll add those also to NETStandard.Library targets file.
